### ISSUE TITLE
Rate limiting implementation for reactor is broken

### ISSUE THREAD

**Issue by thekalinga** [resilience4j/resilience4j#311]
Please ignore this completey, Just look at below comment as it makes more sense than this

~~Rate limiter is supposed to protect a resource against unexpected number of calls like lets say number of times an external API can called/sec~~

~~Meaning, we guard the actual calls, not the results from external call~~

~~In our current Reactor implementation (I assume the same is true for RxJava2 aswell), we are rate limiting the downstream results, not the actual upstream subscription (Refer to `MonoCallable#subscribe`)~~

~~Rate limiter should limit the number of time a method is called, so, for `Mono.fromCallable` the upstream subscription should be delayed (using `delaySubscription`) so that the actual method under rate limit never gets executed till a rate limit token is available~~

~~In the current implementation~~

~~`ResilienceBaseSubscriber#onSubscribe` - is a blind request to upstream without any rate limit
`RateLimiterSubscriber#hookOnNext` - we are rate limiting downstream elements aswell, which is wrong completely, as we should never care about how many elements upstream is producing, we are interested only in rate limiting the upstream calls~~

~~For rate limiting downstream elements within a `Flux`, we have `Flux.limitRate` & `Flux.limitRequest`~~

~~So the current limiter is completely broken~~

**Comment by thekalinga** [comment]
Looks like I'm wrong about half of my analysis. i.e about rate limiting subscription

Looks like the current implementation of `ResilienceBaseSubscriber#isCallPermitted()` is a blocking operation. So all the calls(+ threads) if any of the threads use the same rate limiter will be blocked (blocking thread is not a good idea to begin with) when rate limiter does not have enough tokens to issue

Also, no one asked for limiting downstream events aswell along with upstream subscription

So, the downstream elements are also rate limited for some unknown reason

There are a total of three aspects to rate limit in a flux

1. Intercepting upstream subscription (applicable both to Flux & Mono)
2. Intercepting upstream requests (applicable only to Flux)
3. Intercepting downstream elements (applicable only to Flux)

And each of them should be opt-in. Currently (2) is not at all considered as part of rate limiting, (1) & (3) come in a single package (they are no opt-in), that too by blocking thread

So, the current implementation handles none of the three scenarios correctly

**Comment by storozhukBM** [comment]
Hi @thekalinga,
Thank you for bringing up this conversation. I think we should wait for @madgnome for comment.
If I get your ideas right it should be possible to construct your solution on top of [Metrics.getAvailablePermissions](https://github.com/resilience4j/resilience4j/blob/85107d7acf062e56d74c9b65fcc3288e95f01f85/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java#L361) and [RateLimiter. reservePermission](https://github.com/resilience4j/resilience4j/blob/85107d7acf062e56d74c9b65fcc3288e95f01f85/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiter.java#L281)

Any how if you want to create PR and demonstrate benefits of your solution, it is more than welcome.

**Comment by thekalinga** [comment]
@storozhukBM I'm attempting to modify exiting source to see if I can make the changes alone because

1. I'm not quite certain about why hooks always check for `wasPermitGranted`
2. The a lot of code is common across rate limiter, bulkhead & circuit breaker

I'm attempting to modify it now. Will share an update if I make any progress

**Comment by storozhukBM** [comment]
Sure, thanks.

**Comment by madgnome** [comment]
It seems that you are mixing 2 different concerns in the same issue and labelling everything broken.

## Blocking threads ##

As you have mentioned `RateLimiterSubscriber#isCallPermitted` is a blocking operation.

This is not the most efficient solution but this is not wrong and it is compliant with the reactive stream specification since publishers are not blocked (thanks to `publishOn()` usage).

When I implemented the reactor API, non blocking API for rate limiter was not available. Now that `RateLimiter#reservePermission` is available we may optimize the current solution. I will have a look.

## API ##

The current reactor implementation follows rxjava implementation, it was implemented this way at the time for consistency with the existing API.

As you have said rate limiting can happen in multiple places:
1) subscription (flux and mono)
2) upstream requests
3) elements flowing downstream (for flux)

The current implementation does 1 and 3. 
```
callRemoteService(request)
   .compose(RateLimiterOperator.of(rateLimiter))
```
`callRemoteService` will only be subscribed to if permits are available and elements will only flow if permits are available.

You can achieve something similar to 2 by modifying the chain of operators:

```
Mono.just(request)
    .compose(RateLimiterOperator.of(rateLimiter))
    .flatMap(r -> callRemoteService(r))
```

Here `callRemoteService` will be rate limited.

The current operator makes more sense on `Mono` than it does on `Flux` since we would only check permit once on subscription.

For Flux we could change the implementation (or introduce a new operator) to instead rate limit the requests.

## Side note on open source and leaving feedback ##

You will get better results if you don't start blindly insulting people contribution. You should seek first to understand instead of assuming that people are stupid and only you can save the world.

**Comment by RobWin** [comment]
How do we proceed?

**Comment by madgnome** [comment]
I think we should modify rateLimit to rate limit on the subscription and change the implementation to be non blocking using `reservePermission`

I can raise a PR for it if we agree on the plan.

**Comment by RobWin** [comment]
Yes, a PR would be awesome

**Comment by RobWin** [comment]
I fixed this in an upcoming PR.
Will be solved in 0.15.0
